core/src/block_creation_loop.rs

```diff
@@ -448,43 +570,61 @@ fn shutdown_and_drain_record_receiver(
 /// - Insert the alpentick
 /// - Clear the working bank
 fn record_and_complete_block(
-    poh_recorder: &RwLock<PohRecorder>,
-    record_receiver: &mut RecordReceiver,
-    block_timer: Instant,
+    ctx: &mut LeaderContext,
+    slot: Slot,
+    mut optimistic_parent: Option<(Slot, Hash)>,
+    block_timer: &mut Instant,
     block_timeout: Duration,
 ) -> Result<(), PohRecorderError> {
     loop {
-        let remaining_slot_time = block_timeout.saturating_sub(block_timer.elapsed());
-        if remaining_slot_time.is_zero() {
+        // Check if we've exceeded the block timeout
+        let timeout = time_left(*block_timer, block_timeout);
+        if timeout.is_zero() {
```

Comment:
I think we only break once we receive the ParentReady
```suggestion
        if timeout.is_zero() && optimistic_parent.is_none() {
```

If the cluster is having network issues leading to a slow parent ready we shouldn't punish the leader